### DISCUSSION TITLE
How to avoid double render in case of using useState inside of middleware

### DISCUSSION THREAD

**Discussion by mercs600** [nuxt/nuxt#20470] [Questions]
Hi, I'm wondering how to resolve issue of doubled rendering of components when using global middleware to fetch data and save it on a useState object. 

**Context:**
1. We use dynamic routing and page [...slug].vue
2. We use global route middleware that reads current path, fetches data from API and saves it to page state object which is an implementation of useState
3. We use the  page state object in [...slug].vue to render dynamic components

**Demo and scenario:**
I have reproduced the issue on https://stackblitz.com/edit/github-giyt7k?file=middleware/page.global.ts 
1. Open page on home page
1. Click some product link (this is dynamic page) 
2. You will se doubled console.log('render') 

This happens because of reactivity of useState - we update page state while middleware is being processed, resulting in components being rendered both - before and after the middleware is resolved.

This happens only on the client side when you switch route 

In nuxt2, this issue was easily resolved by using the useAsyncData function in the page component. However, I'm looking for the best way to do this in nuxt3.
For example we could save page state on another callback eg. `router.afterEach` but I'm wondering do you have better option ?

**Comment by ndragun92** [comment]
Looks like something in middleware but U can make it work like this too if it fits your solution

```html
<template>
  <div>
    <NuxtLink to="/">home</NuxtLink>
    <component
      :is="Array.isArray(pageData.data) ? 'ProductList' : 'ProductPage'"
      :data="pageData.data"
    />
  </div>
</template>
<script lang="ts" setup>
const data = usePageState();

const route = useRoute();
const pageData = reactive({
  data: [],
});

const get = async () => {
  try {
    const resp = await $fetch(
      `https://fakestoreapi.com/products/${route.fullPath}`
    );
    pageData.data = resp;
  } catch (err) {
    throw showError({
      statusCode: 404,
      fatal: true,
      unhandled: false,
    });
  }
};

await get();

watch(
  () => route.path,
  async () => {
    await get();
  }
);
</script>

```

**Reply by mercs600** [reply]
yes, it works when you implement it inside the page component, but this is not my case, we need to implement middleware which will be provided by module
